### PR TITLE
Implement support for error messages

### PR DESCRIPTION
**Pull request by aholovko**
These are breaking changes so updating to the next major version is needed here (version 5.0). The reasoning behind changing the API was provided in a comment to the commit "Redesign notifier API".

### CODE REVIEW

File: test/Sharpbrake.Client.Tests/UtilsTests.cs [hunk 1]
```diff
@@ -412,5 +413,34 @@ public void ApplyFilters_ShouldApplyFiltersInOrder()
...
+            {
+                Id = "0005488e-8947-223e-90ca-16fec30b6d72",
+                Url = "https://airbrake.io/locate/0005488e-8947-223e-90ca-16fec30b6d72",
+                Status = RequestStatus.Success
+            };
+
+            Utils.LogResponse(logFile, response);
+
+            Assert.True(File.Exists(logFile));
+            Assert.True(!string.IsNullOrEmpty(File.ReadAllText(logFile)));
```
Comment (kyrylo) [inline test/Sharpbrake.Client.Tests/UtilsTests.cs]: Will `logFile` remain on disk if these assertions fail?

File: test/Sharpbrake.Client.Tests/UtilsTests.cs [hunk 2]
```diff
@@ -412,5 +413,34 @@ public void ApplyFilters_ShouldApplyFiltersInOrder()
...
+
+        [Fact]
+        public void LogResponse_ShouldNotLogIfResponseEmpty()
+        {
+            var logFile = Guid.NewGuid() + ".log";
+
+            Utils.LogResponse(logFile, null);
+
+            Assert.True(!File.Exists(logFile));
+            File.Delete(logFile);
```
Comment (kyrylo) [inline test/Sharpbrake.Client.Tests/UtilsTests.cs]: Will `logFile` remain on disk if these assertions fail?

File: src/Sharpbrake.Client/NoticeBuilder.cs [hunk 3]
```diff
@@ -30,27 +31,50 @@ public NoticeBuilder()
...
-                exception = exception.InnerException;
+
+                ex = ex.InnerException;
+            }
+            else
+            {
+                errors.Add(new ErrorEntry
+                {
+                    Message = message,
+#if !NETSTANDARD1_4
```
Comment (kyrylo) [inline src/Sharpbrake.Client/NoticeBuilder.cs]: Just curious, did NETSTANDARD1_4 not have backtraces (or the StackTrace class)?

File: src/Sharpbrake.Client/NoticeBuilder.cs [hunk 4]
```diff
@@ -30,27 +31,50 @@ public NoticeBuilder()
...
+                    // skip 2 stack frames (for NotifyAsync and SetErrorEntries methods)
+                    Backtrace = Utils.GetBacktrace(new StackTrace(3, true))
+#endif
+                });
+            }
+
+            // to reduce JSON size no more than 3 inner exceptions are processed
+            while (ex != null && errors.Count < 4)
```
Comment (kyrylo) [inline src/Sharpbrake.Client/NoticeBuilder.cs]: I would make `4` a constant and remove the comment:

```c#
private const int MaxInnerExceptions = 3;
```

and then changed this line to:

```c#
while (ex != null && errors.Count <= MaxInnerExceptions)
```

File: src/Sharpbrake.Extensions.Logging/AirbrakeLogger.cs [hunk 5]
```diff
@@ -32,6 +32,9 @@ public AirbrakeLogger(AirbrakeNotifier notifier, IHttpContextAccessor contextAcc
         /// </summary>
         public void Log<TState>(LogLevel logLevel, EventId eventId, TState state, Exception exception, Func<TState, Exception, string> formatter)
         {
+            if (!IsEnabled(logLevel))
```
Comment (kyrylo) [inline src/Sharpbrake.Extensions.Logging/AirbrakeLogger.cs]: Should this be tested?

File: README.md [hunk 6]
```diff
@@ -325,32 +325,24 @@ var config = AirbrakeConfig.Load(settings);
 
 ### AirbrakeNotifier
 
-#### Notify
+#### CreateNotice
```
Comment (kyrylo) [inline README.md]: What do you think about renaming this to `BuildNotice`? We have this method in some of our notifiers, so it should be familiar to old users.

File: test/Sharpbrake.Client.Tests/UtilsTests.cs (hunk 1 above)
Comment (aholovko) [reply test/Sharpbrake.Client.Tests/UtilsTests.cs]: Yes. Fixed. Good catch, thanks.

File: test/Sharpbrake.Client.Tests/UtilsTests.cs (hunk 2 above)
Comment (aholovko) [reply test/Sharpbrake.Client.Tests/UtilsTests.cs]: Fixed.

File: src/Sharpbrake.Client/NoticeBuilder.cs (hunk 3 above)
Comment (aholovko) [reply src/Sharpbrake.Client/NoticeBuilder.cs]: .NET Standard 1.x doesn't implement a lot of APIs available in the standard .NET Framework (so-called "Desktop"), most of them were added later - in [.NET Standard 2.x](https://raw.githubusercontent.com/dotnet/standard/master/docs/versions/netstandard2.0_diff.md) (search for "StackTrace"). For 1.x branch some APIs were implemented in separate packages, for StackTrace it's "System.Diagnostics.StackTrace" (version 4.3.0) - but it adds only the variant with Exception in its constructor that doesn't help as we don't have Exception in that case.

File: src/Sharpbrake.Client/NoticeBuilder.cs (hunk 4 above)
Comment (aholovko) [reply src/Sharpbrake.Client/NoticeBuilder.cs]: Done.

File: src/Sharpbrake.Extensions.Logging/AirbrakeLogger.cs (hunk 5 above)
Comment (aholovko) [reply src/Sharpbrake.Extensions.Logging/AirbrakeLogger.cs]: We don't have tests for integrations.

File: README.md (hunk 6 above)
Comment (aholovko) [reply README.md]: Good idea. Done.

File: src/Sharpbrake.Client/NoticeBuilder.cs [hunk 7]
```diff
@@ -3,16 +3,19 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Runtime.Serialization.Json;
 using System.Text;
 using Sharpbrake.Client.Model;
 
+[assembly: InternalsVisibleTo("Sharpbrake.Client.Tests")]
```
Comment (kyrylo) [inline src/Sharpbrake.Client/NoticeBuilder.cs]: So this class is not accessible to external libraries, but you can access it in tests. That's pretty cool! I haven't seen anything like this for other languages (but probably something like this exists elsewhere, too 😀)

File: docs/migration-guide-from-v4-to-v5.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
+Migration guide from v4 to v5
+=============================
+
+We're happy to announce that Sharpbrake version 5 brings support for error messages along with exceptions. This functionality comes with changes to our API and here are steps to migrate to the new version:
```
Comment (kyrylo) [inline docs/migration-guide-from-v4-to-v5.md]: Would be nice to mention that `(previously, it supported only exceptions)`.

File: docs/migration-guide-from-v4-to-v5.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+1. Use the `BuildNotice` method instead of `NotifyAsync` to pass `exception` object. The `SetHttpContext` method of `notifier` can be used to set `HttpContext`. Call `NotifyAsync` with `notice` returned from `BuildNotice`.
```
Comment (kyrylo) [inline docs/migration-guide-from-v4-to-v5.md]: This will look more friendly:

> To take advantage of the new API, use the `BuildNotice` method instead of....

File: docs/migration-guide-from-v4-to-v5.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+2. The `Notify` method was removed. It was a convenience method around `NotifyAsync` that doesn't bring any real value. The same functionality can be achieved using `NotifyAsync` + continuation:
+
+   ```csharp
+   notifier.NotifyAsync(notice).ContinueWith(responseTask =>
+   {
+       if (responseTask.IsFaulted)
+           Console.WriteLine("Response task is faulted");
+       else
+       {
```
Comment (kyrylo) [inline docs/migration-guide-from-v4-to-v5.md]: Style question: why do we not omit brackets here, but omit them for the line above?

File: docs/migration-guide-from-v4-to-v5.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+           // do something with the Airbrake response
+           var airbrakeResponse = responseTask.Result;
+       }
+   });
+   ```
+
+3. Custom `ILogger` implementation passed to the `AirbrakeNotifier` constructor is not supported anymore. It's unlikely that someone would implement custom ILogger interface for the logger itself so it was removed to simplify API. You still can log responses from Airbrake by specifying `LogFile` in the configuration.
```
Comment (kyrylo) [inline docs/migration-guide-from-v4-to-v5.md]: Missing *the* 
> it was removed to simplify **the** API.

File: src/Sharpbrake.Client/Utils.cs [hunk 12]
```diff
@@ -204,6 +206,54 @@ public static IList<Regex> CompileRegex(IList<string> stringPatterns)
...
+        /// Gets message with properties positionally formatted into the message template
+        /// using culture-specific formatting information.
+        /// </summary>
+        /// <param name="provider">Object that supplies culture-specific formatting information.</param>
+        /// <param name="messageTemplate">Message template that represents .NET composite format string.</param>
+        /// <param name="propertyValues">Zero or more properties to format.</param>
+        /// <returns></returns>
+        public static string GetMessage(IFormatProvider provider, string messageTemplate, params object[] propertyValues)
+        {
+            // TODO: Add support for message templates for structured logging (https://messagetemplates.org/)
```
Comment (kyrylo) [inline src/Sharpbrake.Client/Utils.cs]: I guess this will be addressed in a separate PR? Just pinging you in case you forgot to resolve this TODO (maybe not 😁).

File: docs/migration-guide-from-v4-to-v5.md (hunk 10 above)
Comment (aholovko) [reply docs/migration-guide-from-v4-to-v5.md]: Missing brackets in the first branch implies that this is completed logic (`responseTask` is faulted - nothing can be done with it so just report on the issue and stop) while the second branch (responseTask is Okay) encourage for some additional processing that will span across multiple lines of code ;)

File: src/Sharpbrake.Client/Utils.cs (hunk 12 above)
Comment (aholovko) [reply src/Sharpbrake.Client/Utils.cs]: It's a long story... I spent 2 days just on tokenizer (attach "meaning" to words from the template). And it still failed on edge cases (like [this](https://github.com/serilog/serilog/blob/dev/test/Serilog.Tests/Parsing/MessageTemplateParserTests.cs#L89) or [this](https://github.com/serilog/serilog/blob/dev/test/Serilog.Tests/Parsing/MessageTemplateParserTests.cs#L158)). The best implementation in the field that I found is in Serilog but it's quite complicated and the code just for parsing the message template with properties evaluation is several times bigger than all our codebase including integrations. So I decided to freeze that part for now. Maybe we will return to it in the future. As for now, it requires too many efforts for the value it brings (in our case). I decided to go with the standard .NET format strings (+ string interpolations are supported as well) that with the Airbrake backend do a quite good job.
